adapter/avro/src/main/java/org/apache/arrow/adapter/avro/ArrowToAvroUtils.java

```diff
@@ -0,0 +1,194 @@
+
+    // Avro understands nullable types as a union of type | null
+    // Most nullable fields in a VSR will not be unions, so provide a special wrapper
+    if (nullable && minorType != Types.MinorType.UNION) {
+      final BaseAvroProducer<?> innerProducer = createProducer(vector, false);
+      return new AvroNullableProducer<>(innerProducer);
+    }
+
+    switch (minorType) {
+
+        // Primitive types with direct mapping to Avro
+
+      case NULL:
+        return new AvroNullProducer((NullVector) vector);
+      case BIT:
+        return new AvroBooleanProducer((BitVector) vector);
+      case INT:
+        return new AvroIntProducer((IntVector) vector);
+      case BIGINT:
+        return new AvroLongProducer((BigIntVector) vector);
+      case FLOAT4:
+        return new AvroFloatProducer((Float4Vector) vector);
+      case FLOAT8:
+        return new AvroDoubleProducer((Float8Vector) vector);
+      case VARBINARY:
+        return new AvroBytesProducer((VarBinaryVector) vector);
+      case FIXEDSIZEBINARY:
+        return new AvroFixedProducer((FixedSizeBinaryVector) vector);
+      case VARCHAR:
+        return new AvroStringProducer((VarCharVector) vector);
+
+        // Logical types
+
+      case DECIMAL:
+        return new AvroDecimalProducer.FixedDecimalProducer(
+            (DecimalVector) vector, DecimalVector.TYPE_WIDTH);
+      case DATEDAY:
+        return new AvroDateProducer((DateDayVector) vector);
+      case TIMEMILLI:
+        return new AvroTimeMillisProducer((TimeMilliVector) vector);
+      case TIMEMICRO:
+        return new AvroTimeMicroProducer((TimeMicroVector) vector);
+      case TIMESTAMPMILLI:
+        return new AvroTimestampMillisProducer((TimeStampMilliVector) vector);
+      case TIMESTAMPMICRO:
+        return new AvroTimestampMicroProducer((TimeStampMicroVector) vector);
+
+        // Complex types
+
+      case STRUCT:
+        StructVector structVector = (StructVector) vector;
+        List<FieldVector> childVectors = structVector.getChildrenFromFields();
+        Producer<?>[] childProducers = new Producer<?>[childVectors.size()];
+        for (int i = 0; i < childVectors.size(); i++) {
+          FieldVector childVector = childVectors.get(i);
+          childProducers[i] = createProducer(childVector, childVector.getField().isNullable());
+        }
+        return new AvroStructProducer(structVector, childProducers);
+
+      case LIST:
+        ListVector listVector = (ListVector) vector;
+        FieldVector itemVector = listVector.getDataVector();
+        Producer<?> itemProducer = createProducer(itemVector, itemVector.getField().isNullable());
+        return new AvroArraysProducer(listVector, itemProducer);
+
+      case MAP:
+        MapVector mapVector = (MapVector) vector;
+        StructVector entryVector = (StructVector) mapVector.getDataVector();
+        VarCharVector keyVector = (VarCharVector) entryVector.getChildrenFromFields().get(0);
```

Comment:
What if the key is not a string type?

Reply:
Good point! I put a check in the schema generation but not here. In Avro keys have to be string so if that is not the case I have thrown an error. Please leave this open and I will fix - out of time for today!

Reply:
I added a check for this, now it will throw illegal argument if the key type is not string (which is required by Avro)